Automattic/jetpack
# Sync Checksums :: Convert text to latin1 vs utf8 when performing checksums

Th WP.com cache dbs utilize latin1 as the charset. This leads to known conversion differences when UTF8 data is saved to latin1

FormaÃ§Ã£o de coÃ¡gulo    ->    Formação de coágulo

When we do our checksums comparing text fields this difference in the field data leads to a mismatch in the checksums.

3385248339      vs        3056320041

We previously tried to resolved this issue by having the latin1 data converted to UTF8 but this lead to more discrepancies as the latin1 chars are actually valid utf8 chars and so checksums would not always re-align.  The proper fix for this is that we need to convert the remote field to latin1 and calculate the checksum. When doing so we get a matching hash.

CRC32( convert( 'FormaÃ§Ã£o de coÃ¡gulo' using latin1 ) ) -> 3056320041

This change will not be enabled when this PR is merged and released. It is gated on WP.com and will be explicitly enabled with additional testing and monitoring. 

Please note that there is no db performance hit from this conversion as the convert happens in cpu and is performed as other rows are read. Detailed analysis was done when we thought converting to utf8 would solve our character conversion issue.

Release Note :: Again if we do a point release it would be good to have this included so we can roll it out sooner but 10.3 is totally acceptable as well.

#### Changes proposed in this Pull Request:
* Convert text fields to latin1 to ensure same charset encoding during checksum generation.

#### Does this pull request change what data or activity we track or use?
NO

#### Testing instructions:

* On your test site ensure the db charset is utf8 or utrf8mb4 and add some meta with utf8 supported characters.
* Go to shell
* $store = new Automattic\Jetpack\Sync\Replicastore();
* $store->checksum_all();
* $store->checksum_all( true );

Verify the checksums returned are different for the meta tables where you added utf8 chars.
